### PR TITLE
Support multiple `on parameter set` callbacks #393

### PR DESCRIPTION
**Pull request by suab321321**
@ivanpauno 
Is my method add_on_set_callback correctly implemented?

### CODE REVIEW

**Comment by ivanpauno** [comment]
Please, do not open multiple PRs, always update https://github.com/ros2/rclpy/pull/423.

**Comment by suab321321** [comment]
Okay sir..I m having trouble updating to that same PR..it is always creating its new PR

File: .vscode/c_cpp_properties.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (ivanpauno) [inline .vscode/c_cpp_properties.json]: Delete unrelated file.

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (ivanpauno) [inline .vscode/settings.json]: Delete unrelated file

File: log/latest [hunk 3]
```diff
@@ -0,0 +1 @@
+latest_list
```
Comment (ivanpauno) [inline log/latest]: delete unrelated file

File: log/latest_list [hunk 4]
```diff
@@ -0,0 +1 @@
+list_2019-09-14_15-26-13
```
Comment (ivanpauno) [inline log/latest_list]: this one too

File: rclpy/rclpy/node.py [hunk 5]
```diff
@@ -756,6 +756,20 @@ def _set_parameters_atomically(
             self._parameter_event_publisher.publish(parameter_event)
 
         return result
+    
+    def add_on_set_paramters_callback(self,callback:Callable([List[Parameter],SetParametersResult])):
+        """add the callback to list"""
+        prevCallBack=self._parameters_callback
+        prevCallBack.insert(0,callback)
```
Comment (ivanpauno) [inline rclpy/rclpy/node.py]: you can do:
```python3
self._parameters_callback.insert(0, callback)
```

**Comment by ivanpauno** [comment]
> Okay sir..I m having trouble updating to that same PR..it is always creating its new PR

The PR should be automatically updated when you push to your branch, the diff may not be updated when it has been closed.
In that case, left a comment in the PR asking to be reopen, instead of opening a new one.

**Comment by suab321321** [comment]
Okay @ivanpauno  I got it

**Comment by suab321321** [comment]
@ivanpauno  is this correct?

**Comment by suab321321** [comment]
@ivanpauno reopen

File: .gitignore [hunk 6]
```diff
@@ -56,3 +56,7 @@ docs/_build/
 # PyBuilder
 target/
 .pytest_cache
+
+#vscode extensions
+.vscode
+
```
Comment (ivanpauno) [inline .gitignore]: please, delete this file

**Comment by ivanpauno** [comment]
> @ivanpauno is this correct?

Yes, you still need to modify how the callbacks are called

**Comment by suab321321** [comment]
> > @ivanpauno is this correct?
> 
> Yes, you still need to modify how the callbacks are called

**Comment by suab321321** [comment]
> > @ivanpauno is this correct?
> 
> Yes, you still need to modify how the callbacks are called
@ivanpauno  what is meant by this..
According to me there are two function that will use add_on_set_callback()
1. _set_paramters_atomically
2. _set_parameters
but the comments are showing that these will trigger the callback which is registred with  methods on_set_parameters so where is the role of my implemented methods.

I tried to understanding it from rclcpp but cant seem to figure it out.

File: .gitignore (hunk 6 above)
Comment (suab321321) [reply .gitignore]: .gitignore file was present before also when I git cloned it.I just added my .vscode extension to it..should I delete the .vscode extension or the whole .gitignore

**Comment by suab321321** [comment]
@ivanpauno sir could you please help me

**Comment by suab321321** [comment]
@ivanpauno  I have some doubts

1.I cant find where the callbacks are called so dont know how to use my two functions.
I just little hint sir to complete this,for reference I tried understanding it from rclcpp but cant seem to understand it.

**Comment by suab321321** [comment]
@mabelzhang could you please help me

File: rclpy/rclpy/node.py [hunk 7]
```diff
@@ -141,7 +141,7 @@ def __init__(
         self.__guards: List[GuardCondition] = []
         self.__waitables: List[Waitable] = []
         self._default_callback_group = MutuallyExclusiveCallbackGroup()
-        self._parameters_callback = None
+        self._parameters_callback:Callable[List[Parameter],SetParametersResult]=[]
```
Comment (jacobperron) [inline rclpy/rclpy/node.py]: Thanks for adding a type annotation, you should add some spaces to please the linters and make it more readable (similar to the other annotations above):

 ```suggestion
        self._parameters_callback: Callable[List[Parameter], SetParametersResult] = []
```

File: rclpy/rclpy/node.py [hunk 8]
```diff
@@ -756,6 +756,22 @@ def _set_parameters_atomically(
             self._parameter_event_publisher.publish(parameter_event)
 
         return result
+    
+    def add_on_set_paramters_callback(self,callback:Callable([List[Parameter],SetParametersResult])):
+        """add the callback to list"""
+
+        prevCallBack = self._parameters_callback
+        prevCallBack.insert(0,callback)
```
Comment (jacobperron) [inline rclpy/rclpy/node.py]: As @ivanpauno suggested, you can do the insert on `self._parameters_callback` instead of using a temporary variable.

File: rclpy/rclpy/node.py [hunk 9]
```diff
@@ -756,6 +756,22 @@ def _set_parameters_atomically(
             self._parameter_event_publisher.publish(parameter_event)
 
         return result
+    
+    def add_on_set_paramters_callback(self,callback:Callable([List[Parameter],SetParametersResult])):
```
Comment (jacobperron) [inline rclpy/rclpy/node.py]: Please run linter tests, it will catch errors like this, also, I think you need to use square brackets (`[` `]`) in type annotations  instead of parentheses:
```suggestion
    def add_on_set_paramters_callback(self, callback: Callable[[List[Parameter], SetParametersResult]]):
```

File: rclpy/rclpy/node.py [hunk 10]
```diff
@@ -756,6 +756,22 @@ def _set_parameters_atomically(
...
+        """add the callback to list"""
+
+        prevCallBack = self._parameters_callback
+        prevCallBack.insert(0,callback)
+        self._parameters_callback = prevCallBack
+
+    
+    def remove_on_set_parameters_callback(self,callback:Callable[List[Parameter],SetParametersResult]):
```
Comment (jacobperron) [inline rclpy/rclpy/node.py]: Fix linter errors here.

File: rclpy/rclpy/node.py [hunk 11]
```diff
@@ -756,6 +756,22 @@ def _set_parameters_atomically(
...
+        """remove callback from list"""
+        try:
+            self._parameters_callback.remove(callback)
+        except ValueError:
+            print("Callback doesn't exist")
```
Comment (jacobperron) [inline rclpy/rclpy/node.py]: I guess it's okay to let the exception go (ie. remove the `try-except`), instead of catching it and printing an error message.

File: rclpy/rclpy/node.py (hunk 7 above)
Comment (suab321321) [reply rclpy/rclpy/node.py]: If I change it in your way then I wont be able to add element into the self._parameters_callback it will instead override the self._parameters_callback as far I see it

**Comment by suab321321** [comment]
@jacobperron 
 If I change it in your way then I wont be able to add element into the self._parameters_callback it will instead override the self._parameters_callback as far I see it

**Comment by jacobperron** [comment]
> If I change it in your way then I wont be able to add element into the self._parameters_callback it will instead override the self._parameters_callback as far I see it

```python
self._parameters_callback.insert(0, callback)
```
is not overriding anything, it is modifying the list by inserting a new element at index 0.

**Comment by suab321321** [comment]
Okay sir will update soon

**Comment by suab321321** [comment]
Why are my pull request not passing successfully

**Comment by jacobperron** [comment]
@suab321321 Looking at the CI log, it looks like there's a formatting error:

```
E         File "/tmp/ws/src/rclpy/rclpy/rclpy/node.py", line 783
E           if callback in self._parameters_callback:
E                                                   ^
E       IndentationError: unindent does not match any outer indentation level
```

Before pushing changes, it's a good idea to test changes locally. This helps us in the review process.
You should be able to reproduce the test errors; if you are using colcon:

    colcon test --packages-select rclpy
    colcon test-result --verbose

**Comment by suab321321** [comment]
> @suab321321 Looking at the CI log, it looks like there's a formatting error:
> 
> ```
> E         File "/tmp/ws/src/rclpy/rclpy/rclpy/node.py", line 783
> E           if callback in self._parameters_callback:
> E                                                   ^
> E       IndentationError: unindent does not match any outer indentation level
> ```
> 
> Before pushing changes, it's a good idea to test changes locally. This helps us in the review process.
> You should be able to reproduce the test errors; if you are using colcon:
> 
> ```
> colcon test --packages-select rclpy
> colcon test-result --verbose
> ```

Okay sir,I will update soon

**Comment by suab321321** [comment]
@jacobperron sir I can understand what continuation line unaligned for hanging indent:
            callback: Callable[[List[Parameter]], SetParametersResult]
this error is..this is as
 def set_parameters_callback(
        self,
        callback: Callable[[List[Parameter]], SetParametersResult]
    ) -> None:
but it is showing error in method....
And also it is showing error on multiple file like client.py,action_graph.py and others but I did not touch the files so how come those are showing errors

And I have made this changes how in callbacks are called

 for parameter in parameter_list:
                result = self._parameters_callback(parameter)

File: install/.colcon_install_layout [hunk 12]
```diff
@@ -0,0 +1 @@
+isolated
```
Comment (allenh1) [inline install/.colcon_install_layout]: @suab321321 Did you run colcon inside the `rclpy` folder? You should probably run it from the root of the workspace.

Also you might want to clean up the `install`, `log`, `build` folders.

```
git rm -rf build install log
```

Then commit the result and push it.

Comment (suab321321) [reply install/.colcon_install_layout]: @allenh1  Sir my main problem is when 
`
for callback in self._parameters_callback:
                result = callback(parameter_list)
                if(not result.successful):
                    return result
`
this is showing raising TypeError method is not iteratable.I tried to find the error in stackoverflow but cannot understanding why it is failing here in this case

Comment (allenh1) [reply install/.colcon_install_layout]: What is the type of `self._parameters_callback`? This error seems to imply that `self._parameters_callback` is a method, _not_ an iterable (such as a list of methods).

But if you could clean up the diff (by removing all the extra files), it would be a lot easier for people to help you fix the issues (since there's fewer changes to look at).

Comment (suab321) [reply install/.colcon_install_layout]: - Its type is Callable[[List[Parameter],
SetParameterResult]

I have deleted the log build and install folders,and I have no idea how this file changed I am working with node.py

Should I reinitialize my local repo again with the master?

Comment (allenh1) [reply install/.colcon_install_layout]: @suab321321 I think you cloned the `rclpy` repo inside this repo and added it.

> Callable[[List[Parameter],
SetParameterResult]

This is not iterable. You want a list of callables, not a callable (which is why the interpreter is upset).

Comment (suab321321) [reply install/.colcon_install_layout]: @allenh1 I m creating a repo and I will be doing new PR from that.

**Comment by suab321321** [comment]
Continued in Support multiple parameters callback #457
